clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java

```diff
@@ -31,32 +31,155 @@
+                // Increment the partition count as we have encountered a new partition
+                partitionCount++;
+                // Update the current partition to the new partition
+                currentPartition = record.partition();
+            }
+
+            validateRecordPayload(topic, record, currentPartition, recordCount, recordSize);
+            recordCount++;
+        }
+
+        // Including empty partition
+        assertEquals(partitionSize, partitionCount + 1);
+    }
+
+    @Test
+    public void testRecordsByPartition() {
+        List<String> topics = Arrays.asList("topic1", "topic2");
+        int recordSize = 3;
+        int partitionSize = 5;
+        int emptyPartitionIndex = 2;
+
+        ConsumerRecords<Integer, String> consumerRecords = buildTopicTestRecords(recordSize, partitionSize, emptyPartitionIndex, topics);
+
+        for (String topic : topics) {
+            for (int partition = 0; partition < partitionSize; partition++) {
+                TopicPartition topicPartition = new TopicPartition(topic, partition);
+                List<ConsumerRecord<Integer, String>> records = consumerRecords.records(topicPartition);
+
+                if (partition == emptyPartitionIndex) {
+                    assertTrue(records.isEmpty());
+                } else {
+                    assertEquals(recordSize, records.size());
+                    for (int i = 0; i < records.size(); i++) {
+                        ConsumerRecord<Integer, String> record = records.get(i);
+                        validateRecordPayload(topic, record, partition, i, recordSize);
+                    }
+                }
+            }
         }
-        assertEquals(2, c);
+    }
+
+    @Test
+    public void testRecordsByNullTopic() {
+        String nullTopic = null;
+        ConsumerRecords<Integer, String> consumerRecords = ConsumerRecords.empty();
+        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> consumerRecords.records(nullTopic));
+        assertEquals("Topic must be non-null.", exception.getMessage());
+    }
+
+
+    @Test
+    public void testRecordsByTopic() {
+        List<String> topics = Arrays.asList("topic1", "topic2", "topic3", "topic4");
+        int recordSize = 3;
+        int partitionSize = 10;
+        int emptyPartitionIndex = 6;
+        int expectedTotalRecordSizeOfEachTopic = recordSize * (partitionSize - 1);
+
+        ConsumerRecords<Integer, String> consumerRecords = buildTopicTestRecords(recordSize, partitionSize, emptyPartitionIndex, topics);
+
+        for (String topic : topics) {
+            Iterable<ConsumerRecord<Integer, String>> records = consumerRecords.records(topic);
+            Iterator<ConsumerRecord<Integer, String>> iterator = records.iterator();
+            int recordCount = 0;
+            int partitionCount = 0;
+            int currentPartition = -1;
+
+            while (iterator.hasNext()) {
```

Comment:
Maybe we can loop `records`? `for (ConsumerRecord<Integer, String> record : records) {`

Reply:
Indeed, it will be more easy to read than using iterator.